server/fleet/datastore.go

```diff
@@ -411,6 +411,11 @@ type Datastore interface {
 	// associated with the given name, version, arch, and kernel version,
 	// a new record is also created.
 	UpdateHostOperatingSystem(ctx context.Context, hostID uint, hostOS OperatingSystem) error
+	// CleanupHostOperatingSystems removes records from the host_operating_system table that are
+	// associated with any non-existent host (e.g., expired hosts) and removes records from the
+	// operating_systems table that no longer associated with any host (e.g., all hosts have
+	// upgraded from a prior version).
+	CleanupHostOperatingSystems(ctx context.Context) error
```

Comment:
(Related) Also, don't forget to add `host_operating_system` to https://github.com/fleetdm/fleet/blob/main/server/datastore/mysql/hosts.go#L276-L292 (so that it gets deleted in `ds.DeleteHost`).